### DISCUSSION TITLE
Shouts out from librespot!

### DISCUSSION THREAD

**Discussion by roderickvd** [jpochyla/psst#204] [General]
Hi Jan!

Congratulations on this project. As one of the `librespot` maintainers, it was recently brought to my attention over at our [Gitter chat](https://gitter.im/librespot-org/spotify-connect-resources). Your implemention looks great and your progress seems impressive.

Next to offering my compliments, I'd like to reach out and see where we can work together. Particularly since we are both working on Rust implementations, it'd be great to leverage the power from combined efforts.

I understand that you have so far done away with the `async` parts but still think that there's a lot going here, particularly in the audio and protocol departments. On our side, we have been doing a lot of improvements in the audio side (dithering and dynamic limiting stand out), and it'd be pretty easy to merge your Symphonia efforts which we've also been eying for MP3 and future FLAC support. On your side, the "new CDN API" is exactly what we've been meaning to implement in our somewhat understaffed [`new-api`](https://github.com/librespot-org/librespot/tree/new-api) branch.

I have no doubts that working together we can do more and the `async` parts will only need to be a minor hindrance in a common code base.

Looking forward to hearing from you.

Cheers on behalf of the rest of the team,

Roderick

**Comment by jpochyla** [comment]
Hi Roderick!

Thanks a lot for the kind words, and for reaching out. I'm also extremely sorry for being such a slowpoke, the audio backend rewrite had been a struggle recently, I pushed everything else aside.

Psst started as a way to kill time in the quarantine, but I still feel a bit guilty for choosing to rewrite everything instead of contributing to `librespot`. It's been, of course, the biggest inspiration, and I'm very grateful for the project. It would be great if we could overcome the sync-async split and share some bits of the code, the audio-related stuff is indeed impressive! I'm also trying to document everything the best I can, and hope that the code can be at least useful as a reference.

Best,

Jan

**Reply by roderickvd** [reply]
Hi @jpochyla great to hear from you. I know how it goes when life happens, all part of the open source vibe.

Surely there is no need to feel guilty as we can only be happy that `librespot` is sparking other projects. But indeed if we could join forces on certain areas, well, the prospect of our mutual progress is enticing.

The prime focus of `librespot` has always been the protocol, second the Connect implementation with the player. Those two live in separate threads and hence the `async` keywords, but I do believe that with all the `awaits` sprinkled around, there's a lot sequential events going on. Then of course the downloading of files (particularly for gapless playback) needs to be done in the background. That you could do without the `asyncs` is maybe the best proof of this point.

All new Spotify features require the new HTTP / web sockets based API with CDN. We don't think that Spotify will turn off the "old" API because they have so many legacy devices in the wild, but still, `librespot` is overdue to move to that API, like `librespot-java` has. I think it's amazing that you ported that effort to Rust which in my mind took a large effort around a new session handler.

Of course I could (and will 😆) look at that part of your implementation, but I'd be delighted to take that up with you and your contributors and have one central, up-to-date protocol implementation in Rust. I do understand that it would incur some heavy lifting to get it working in an existing code base, potentially carving it out, and moving to a unified API, but again I have no doubts about the worth of that small investment and the idea of a larger team of maintainers improving it further and further.

Hoping you'd agree and want to chime in, I'd be super happy to discuss how we could further take this on. We could continue on Gitter, meet up over video, anyway I'm game. 🍻
